src/lib/votecounting.ts

```diff
@@ -0,0 +1,140 @@
+import {
+  Candidate,
+  ContestOption,
+  ContestOptionTally,
+  Dictionary,
+  Election,
+  ElectionTally,
+  VotesByPrecinct,
+  VotesDict,
+} from '../config/types'
+
+// the generic write-in candidate to keep count
+const writeInCandidate: Candidate = {
+  id: 'writein',
+  name: 'Write-In',
+  isWriteIn: true,
+}
+
+interface ParseCVRsParams {
+  election: Election
+  CVRsText: string
+}
+
+export function parseCVRs({ election, CVRsText }: ParseCVRsParams) {
+  // CVRs are newline-separated JSON objects
+  const votesByPrecinct: VotesByPrecinct = {}
+
+  CVRsText.split('\n')
+    .filter(el => el) // remove empty lines
+    .map(line => JSON.parse(line))
+    .forEach(CVR => {
+      const vote: VotesDict = {}
+      election.contests.forEach(contest => {
+        if (!CVR[contest.id]) {
+          return
+        }
+
+        if (contest.type === 'yesno') {
+          // the CVR is encoded the same way
+          vote[contest.id] = CVR[contest.id]
+          return
+        }
+
+        if (contest.type === 'candidate') {
+          vote[contest.id] = CVR[contest.id].map((candidateId: string) =>
+            candidateId === 'writein'
+              ? writeInCandidate
+              : contest.candidates.find(c => c.id === candidateId)
+          )
+        }
+      })
+
+      let votes = votesByPrecinct[CVR._precinctId]
+      if (!votes) {
+        votesByPrecinct[CVR._precinctId] = votes = []
+      }
+
+      votes.push(vote)
```

Comment:
nit: continuing from above comment, wondering if this would make more sense as…
```suggestion
      ballots.push(ballot)
```